numpy/lib/histograms.py

```diff
@@ -782,7 +782,7 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
     
     a, weights = _ravel_and_check_weights(a, weights)
 
-    if not np.issubdtype(a.dtype, np.number):
+    if (a.dtype == '<U21') or (a.dtype == 'S21'):
```

Comment:
Those are very specific to arrays where the maximum string length is 21, so that's not the right check. It might take a little discussion on the parent issue to figure out exactly the set of dtypes that we want to forbid and allow here. It will look like the `np.issubdtype()` check that you just had, but with several more clauses (we want to allow integers and real floats but not `complexfloating`, and we want to allow `bool_` and possibly `datetime64`, I'm not sure).

Reply:
Yes, we will need datetime as well. In test it is failing there.

Reply:
Bool is giving error "NameError: name 'warnings' is not defined" in _ravel_and_check_weights function